sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java

```diff
@@ -106,62 +107,94 @@ public RetryPolicy(RetryOptions retryOptions) {
 
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-        return attemptAsync(context, next, context.getHttpRequest(), 0, null);
+        // Create a buffered version of the request that will be used each retry.
+        // The buffering is done here as once the request body has been buffered once it doesn't need to be buffered
+        // again as it will be buffered as a read-only buffer.
+        HttpRequest originalHttpRequest = context.getHttpRequest();
+        BinaryData originalRequestBody = originalHttpRequest.getBodyAsBinaryData();
+        if (retryStrategy.getMaxRetries() > 0 && originalRequestBody != null && !originalRequestBody.isReplayable()) {
+            return originalRequestBody.toReplayableBinaryDataAsync().flatMap(bufferedBody -> {
```

Comment:
nit:
```suggestion
            return originalRequestBody.toReplayableBinaryDataAsync().flatMap(replayableBody -> {
```

super-minor, but it initially gave me an impression that it's really buffered even if replayable and there is no way to prevent buffering